org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/services/GenericEndpoint.java

```diff
@@ -63,8 +64,10 @@ protected void recursiveFindRpcMethods(Object current, Set<Class<?>> visited, Se
 					Method method = methodInfo.method;
 					Object[] arguments = this.getArguments(method, arg);
 					return (CompletableFuture<Object>) method.invoke(current, arguments);
-				} catch (InvocationTargetException | IllegalAccessException e) {
-					throw new RuntimeException(e);
+				} catch (InvocationTargetException e) {
+					throw new CompletionException(e.getCause());
+				} catch (IllegalAccessException e) {
+					throw new CompletionException(e);
 				}
```

Comment:
I have some second thoughts about this change.

The reason is that this code is used for handling both requests and notifications.

While using `CompletionException` as a wrapper makes sense for requests, it seems much less natural for notifications.

In fact, any exceptions thrown from notification handlers are logged at

https://github.com/eclipse-lsp4j/lsp4j/blob/88868d3506d317b12cf0bfd1e82a254489f8c47b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/RemoteEndpoint.java#L233

and discarded, since notifications are not expected to send anything back. In our case, the logged exception would be the wrapping `CompletionException`, which might seem a bit confusing. We could, of course, expect the `CompletionException` and unwrap it before logging, but it looks more complicated than necessary and odd.

So, I wonder: would using sneaky throw instead of throwing `CompletionException` a more consistent approach for both requests and notifications?

Reply:
I don't have a strong opinion on this - the important thing is that a `ResponseError` gets unwrapped. I don't know how important the intermediary exceptions are.

As of now I have never used sneaky throws, so don't know if they would be suitable. If you think they are, I'm very ok to defer to your experience.

From https://github.com/eclipse-lsp4j/lsp4j/pull/809#discussion_r1491277086:

> [...] if someone want to take this up in the meantime that would be great because my allocated LSP4J time is running out quick.

I can probably find a bit more time, but I hope that someone else can help this come to a conclusion :-)

Reply:
I'll try the sneaky throw approach and see how it goes then. I agree that it probably does not matter much, provided that the main goal of this PR is achieved, but it might be worthwhile to try different approaches, especially since we now have great test coverage around this code area -- thank you :-)

Reply:
In my opinion sneaky throws looks like a hack and seems to work by accident in java.

Reply:
> In my opinion sneaky throws looks like a hack and seems to work by accident in java.

This looks like a strong opinion. But are you speaking in general or do you see any specific drawbacks in using it in this particular case?

Reply:
No one expects a checked exception if it is not specified.
We can simply wrap the exception in a specific `RuntimeException` and then unwrap it before calling the exception handler. This way we can hide the technical detail that we need to wrap it internally.

Reply:
> No one expects a checked exception if it is not specified.

As far as I can see, the exception handler expects a `Throwable` as its argument, which includes any checked 
exception.

Also, looking at the implementation of `RemoteEndpoint.handleRequest` and `handleNotification`, I cannot see any drawbacks with using sneaky throws in this case.

> We can simply wrap the exception in a specific RuntimeException and then unwrap it before calling the exception handler. This way we can hide the technical detail that we need to wrap it internally.

Then, we would need to make the `RemoteEndpoint` aware of this specific runtime exception type used in the `GenericEndpoint`, which does not look like a good design to me.

Reply:
Overall, I agree with you that sneaky throws is not a general purpose tool and should not be used in ordinary code. But for the right job, in essentially reflective code, it can be a valuable tool. It is just based on how the JVM works, not on how the Java compiler works.

Reply:
I agree. May be checked exceptions are anyway very rare. So if we only wrap in that case it would may be also ok?

```
if (e.getCause() instanceof RuntimeException) {
  throw (RuntimeException) e.getCause();
}
throw new CompletionException(e.getCause());
```

Reply:
Sure, this would also be a valid approach. It is just that I'm looking for something that would work a bit more uniformly/consistently across requests/notifications, checked/unchecked exceptions. I have just tried the sneaky throws approach locally and it seems to deliver what I'm looking for. In my experience, it is the right approach in this case. Anyway, thank you for discussing this and other alternatives.